### PR TITLE
Add KarhunenLoeveValidation

### PR DESCRIPTION
**Pull request by jschueller**
- Add a method to draw KLResult eigenvalues
- Add KarhunenLoeveReduction helper class to perform projection + lifting of a field (it's a field->field function):
```
algo = ot.KarhunenLoeveSVDAlgorithm(processSample, threshold)
algo.run()
klresult = algo.getResult()
# Create the KL reduction
meanField = processSample.computeMean()
klreduce = ot.KarhunenLoeveReduction(klresult)
# Generate a trajectory and reduce it
field = process.getRealization()
values = field.getValues()
reducedValues = klreduce(values)
ott.assert_almost_equal(values, reducedValues)
```

- Add KarhunenLoeveValidation class for various graphical validation of KL decomposition (similar to MMValidation):

```
import openturns as ot
N = 20
interval = ot.Interval(-1.0, 1.0)
mesh = ot.IntervalMesher([N - 1]).build(interval)
covariance = ot.SquaredExponential()
process = ot.GaussianProcess(covariance, mesh)
sampleSize = 100
processSample = process.getSample(sampleSize)
threshold = 1.0e-7
algo = ot.KarhunenLoeveSVDAlgorithm(processSample, threshold)
algo.run()
klresult = algo.getResult()
validation = ot.KarhunenLoeveValidation(processSample, klresult)
residualProcessSample = validation.computeResidual()
residualSigmaField = validation.computeResidualStandardDeviation()
residualMean = validation.computeResidualMean()
graph0 = validation.drawValidation()
graph1 = residualProcessSample.drawMarginal(0)
graph2 = residualMean.drawMarginal(0)
graph3 = residualSigmaField.drawMarginal(0)
graph4 = validation.drawObservationWeight(0)
graph5 = validation.drawObservationQuality(5)
```

validation1
validation1-residual
validation1-residual-mean

obsw

validation1-indiv-weight

obsq
validation1-indiv-quality

### CODE REVIEW

File: lib/src/Base/Stat/ProcessSampleImplementation.cxx [hunk 1]
```diff
@@ -195,6 +196,29 @@ Sample ProcessSampleImplementation::computeSpatialMean() const
...
+  for (UnsignedInteger i = 0; i < size; ++i)
+  {
+    Sample slack(data_[i] - meanValues);
+    for (UnsignedInteger j = 0; j < verticesNumber; ++ j)
+      for (UnsignedInteger k = 0; k < dimension; ++ k)
+        slack(j,k) = slack(j, k) * slack(j, k);
+    stdValues += slack;
+  }
+  stdValues *= Point(getDimension(), 1.0 / (size - 1.0));
+  return Field(mesh_, stdValues);
```
Comment (regislebrun) [inline lib/src/Base/Stat/ProcessSampleImplementation.cxx:218]: No square root anywhere?

**Review by regislebrun — CHANGES_REQUESTED** [review]
I couldn't resist to have a look at your code, even if it is mark as 'draft'. Nice work, but I found some glitches you may want to fix

File: lib/src/Base/Algo/KarhunenLoeveValidation.cxx [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+  if (K >= N)
+    throw InvalidArgumentException(HERE) << "K (" << K << ") should be less than the number of modes (" << N << ")";
+  Sample xi(result_.project(sample_));
+  const UnsignedInteger size = xi.getSize();
+  Point w(size);
+  Point wT(size);
+  for (UnsignedInteger i = 0; i < size; ++ i)
+  {
+    for (UnsignedInteger k = 0; k < K; ++ k)
+      w[i] += xi(i, k) * xi(i, k);
```
[the remark this answers is not in the document]
Comment (jschueller) [reply lib/src/Base/Algo/KarhunenLoeveValidation.cxx:165]: yes, also I dont need wTi

File: lib/src/Base/Stat/ProcessSampleImplementation.cxx (hunk 1 above)
Comment (jschueller) [reply lib/src/Base/Stat/ProcessSampleImplementation.cxx:220]: indeed :{

**Comment by regislebrun** [comment]
With the current implementation, one try to build an approximation of the identity function over the space of trajectories of the underlying process using a basis built using the covariance function only. If the resulting basis is not well suited to approximate the mean function, then the approximation will be poor. See the following example:
```
from openturns import *

N = 100
M = 10000
P = 10
mean = SymbolicFunction("x", "sign(x)")
cov = SquaredExponential([1.0], [0.1])
mesh = IntervalMesher([N]).build(Interval(-2.0, 2.0))
process =GaussianProcess(TrendTransform(mean, mesh), cov, mesh) 
sample = process.getSample(M)
algo = KarhunenLoeveSVDAlgorithm(sample, 1e-6)
algo.run()
result = algo.getResult()
metamodel1 = FieldToFieldConnection(KarhunenLoeveLifting(result), KarhunenLoeveProjection(result))
trend = TrendTransform(Function(P1LagrangeEvaluation(sample.computeMean())), mesh)
metamodel2 = FieldToFieldConnection(trend, FieldToFieldConnection(metamodel1, trend.getInverse()))
sample = process.getSample(P)
reduced1 = metamodel1(sample)
reduced2 = metamodel2(sample)
g = sample.drawMarginal(0)
g.setColors(["red"])
g1 = reduced1.drawMarginal(0)
g1.setColors(["blue"])
drs = g1.getDrawables()
for i, d in enumerate(drs):
    d.setLineStyle("dashed")
    drs[i] = d
g1.setDrawables(drs)
g.add(g1)
g2 = reduced2.drawMarginal(0)
g2.setColors(["green"])
drs = g2.getDrawables()
for i, d in enumerate(drs):
    d.setLineStyle("dotted")
    drs[i] = d
g2.setDrawables(drs)
g.add(g2)
Show(g)
```
produces the following picture:
Figure_1

**Comment by jschueller** [comment]
I modified the norm formula, and added an optional trend argument to KL reduction, so this should be good to go.

File: lib/src/Base/Stat/ProcessSampleImplementation.cxx [hunk 3]
```diff
@@ -194,6 +204,32 @@ Sample ProcessSampleImplementation::computeSpatialMean() const
   return result;
 }
 
+/* Standard deviation accessor */
+Field ProcessSampleImplementation::computeStandardDeviation() const
+{
+  const UnsignedInteger size = getSize();
+  if (size == 0) return Field();
+  if (size == 1) return Field(mesh_, getDimension());
+  Sample meanValues(computeMean().getValues());
```
Comment (regislebrun) [inline lib/src/Base/Stat/ProcessSampleImplementation.cxx:213]: this way you go twice through the data. You can compute the marginal variance incrementally then takes the square root of these values, in the spirit of the Monte Carlo update. It should be mush more efficient on large data sets.

**Review by regislebrun — APPROVED** [review]
Nice work. Only minor remarks, excepted the computation of standard deviations which could be much more efficient

File: lib/src/Base/Algo/KarhunenLoeveResultImplementation.cxx [hunk 4]
```diff
@@ -84,11 +84,51 @@ CovarianceModel KarhunenLoeveResultImplementation::getCovarianceModel() const
...
+{
+  Graph graph("Karhunen-Loeve eigenvalues", "Index", "Eigenvalue", true, "topright");
+  const UnsignedInteger K = eigenvalues_.getSize();
+  Point indices(K);
+  for (UnsignedInteger i = 0; i < K; ++ i)
+    indices[i] = i;
+  Cloud cloud(indices, eigenvalues_);
+  graph.add(cloud);
+  graph.setGrid(true);
+  return graph;
```
[the remark this answers is not in the document]
Comment (jschueller) [reply lib/src/Base/Algo/KarhunenLoeveResultImplementation.cxx:108]: that's an option of the graph that the user can set, we actually added a bool to this method but removed it at is built-in in the Graph object anyway, so the user choses here

File: lib/src/Base/Stat/ProcessSampleImplementation.cxx (hunk 3 above)
Comment (jschueller) [reply lib/src/Base/Stat/ProcessSampleImplementation.cxx:213]: it's a tradeoff, because you also have to store the incremental mean estimate

**Pull request merged**